### PR TITLE
GUI Spells, New CD Rotation and GUI Description on spells.

### PR DESCRIPTION
**Pull request by Myrilandel**
## New Features
**GUISpellDisplayer**
- Script associated to the SpellBarPrefab.
- One instance of this script is used for each spell.
- This script allows to display all informations you need to know about a spell ingame. 
- Adding Cooldown informations.
- Adding Description while the cursor is over the spell.
- Adding images per spell.
- Adding Layer to know if a spell is launchable or not in special conditions.
- Adding a text field to display the number of stacks of stackable spells

**SpellData and SpellData.json**
- Serializable Class at the service of Spell class
- Contains every element we need to build a spell.

**StringHelper**
- Static class used to help us ine string construction.
- Has one method to build Spell Description that will be displayed on the GUI.
- Has one method to build friendly timers that are > 60 seconds with a format of mm.s
- Allows us to seperate, magical and physical damages and other value.

## Enhancement
**Character**
- Now able to find an instance of GUISpellDisplayer and associates a spell to it.

**Spell**
- Now has a description field.
- Now get every elements to construct itself from a JSON stream input.
- Now has an abstract method used by inherited classes to fill the description field.
- All inherited scripts has been modified in order to fit with the new abstract method.
- Now has a new cooldown rotation (the spell is ready when the CurrentCD reaches 0).

**FireBall**
- Does not apply damages by itself
- Now notify the FireBallSpell to apply damages.

**SolarBall**
- Does not apply damages by itself
- Now notify the SolarBurnSpell to apply damages at the impact.
- Now notify the SolarBurnSpell to apply damages at the explosion (AOE) for each target.

## KB Tasks & PlantUML
- [KB Task #137](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=137&project_id=3)
- [KB Task #138](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=138&project_id=3)
- [KB Task #139](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=139&project_id=3)
- [KB Task #141](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=141&project_id=3)
- [KB Task #143](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=143&project_id=3)
- [KB Task #146](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=146&project_id=3)
- [KB Task #150](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=150&project_id=3)
- [KB Task #155](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=155&project_id=3)
- [KB Task #156](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=156&project_id=3)

- [UML](http://www.plantuml.com/plantuml/svg/fLLDRziu4BthLmoGYmrN1lPQBnPmhIJrOfCSdFGI10jA7DbiA58WHzvcW_xt7PAI8wNfeaLpYVdcU-RDg1NLezWZNDvjuJdBAYsy3uzTYreNohTQFA5Z4V1V7b7usuT_gmJTuXVgaV2pG5yvrPAo9idlT-YKqEfVl52evGDicn0lMMoR_SXWzVyMcsANRFYr-W7dwwDGMfGQBwpZYhvdaWeWrQ3R5E6nogBcucYLX3MHKsL766FVMdOqIrs8CUOZrIlRUJmtX6xsBVn5KO_c4TsHS5FahfM2C41ViwzPrlQ3kR8IDVTVb9wSg0ZIY68EprcUmr5mFLobXppFpg2qjizjQ-uy1etMsHOTgQIogBKLb5GsnV2iEkVGK03Iw0M95AJ0MXaL1yg-6gI3bT7HSWaNdQc2n0WDLxrakSmMH-MeOomMzfTWhKFqnTKjKePRlplboEZPn69epXGzDMWAtvZeAKL16UKnh1NoLCgCcwVC7ZvxQ-OXyy3tjVUsKknJlfbypFCLYK81lQjB79CzoSvEKxBbPwpe_W7s4xLFmx1cyrL8ey2osmDn959j39-IxpSlvdzjgIykM4xPzwtsng48aQPwFBc1zy49R0A94X_kTq-Ui5cEvG_POaIpqC10oSr1D2sxXbnZJVaYIU64HtZnas2MdMBykRkvpdlq5q7Ezjnq3Ko4yOeDuzl5aLoB9jwI0NtzldrgqpjjoOO7LTX_U8owItWaoKVXBpT5V2j3y45uBCHmfWB4F5XB6PT8w2JoKRPbgbFd9_BpkcRg-54QBBwX0xeOzfLWSZ-vx5TtuZIpvo0qbX24X-lVsCSwaGCgOQ3ib2PGv57NG0TX7Zspad_Xj8LBjjkoeeT88rRXut9K4g4UdGBKsF0nYIxOvoa9_trpPMhhciXS3vnkYCmx5GpMh6rWMjSxzWjDvmbA1BxPQ2JAwDrmdrHOwGQR4frFMaCOyKEB2N1xu1pJhKiE75BdJ4p4EV44jeQRw0a4aGY7P-YNV74AivQdH3LezOZpPSPPGXZ1otM_8wKzpwRVjZZKZwZR-Cdb8ND3AXWOcwxnoVi7dfdaZyreqTzbB_N9DugRP_7RaYpoDxxpx_bCS98NSHPSqoWS_063OJxDzFsWtwjJ9DplBzj-Vr1Iedd89rpw3m00)

### CODE REVIEW

**Comment by Myrilandel** [comment]
Je viens de m'apercevoir que j'ai zappé certaines méthodes au niveau des commentaires, je les ajoute dans une prochain commit aujourd'hui !

EDIT : Désolé pour le close/Re-Open de la PR j'ai miss click :D

**Comment by Myrilandel** [comment]
J'ai ajouté de nouveaux éléments notamment la construction des spells directement depuis un JSON. Ça fait beaucoup d'éléments nouveaux à prendre en main hésitez pas à vous envoyer sur les questions !

**Comment by Veltouille** [comment]
Je commence ma review et déjà l'UML est excellent avec le Model et la description de l'utilisation du JSON.

File: Assets/StreamingAssets/SpellData.json [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+		"NumberOfStack":0,
+		"Description":
+		[
+			"Invokes an enormous eruption from the sky that explodes at the targeted point.\n",
+			"At the impact, if the Solar Burn hits an entity, it sudden {m0} damages.\n",
+			"Otherwise, every target caught in the explosion sudden {m1} damages.\n",
+			"Critical hits inflicts 4x more damages instande of 2x."
+		]
+	}
+]
```
Comment (Veltouille) [inline Assets/StreamingAssets/SpellData.json:77]: ### Question

Ok, donc c'est le JSON qui décrit les sorts d'un champion.
Je vois que seul le FireMage est renseigné, ma question c'est est que ce JSON remplace l'ancien CharacterData.json ou les deux sont nécessaires à la création d'un champion ?

**Comment by Myrilandel** [comment]
Merci j'ai essayé d'être le plus clair possible car le mécanisme n'est pas évident.
tu noteras que j'ai représenté le JSON au format de class Enum mais c'est bel et bien un fichier .JSON (j'avais pas trop le choix pour le représenter).

File: Assets/Scripts/Utils/StringHelper.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        MatchCollection matches = Regex.Matches(description, pattern, RegexOptions.IgnoreCase);
+
+        if (descriptionVariables != null)
+        {
+            for (int i = 0; i < descriptionVariables.Length; i++)
+            {
+                if (matches[i].Value[0] == 'm' || matches[i].Value[0] == 'p')
+                {
+                    string color = matches[i].Value[0] == 'm' ? "cyan" : "maroon";
+                    description = description.Replace("{" + matches[i].Value + "}", "<b><color=" + color + ">{" + matches[i].Value.TrimStart('m') + "}</color></b>");
```
Comment (Veltouille) [inline Assets/Scripts/Utils/StringHelper.cs]: ### Question

J'ai compris que tu fais sauter le 'm' quand tu match sur le pattern et quand tu change la couleur par rapport au type de spell. Mais tu fais sauté le 'p' aussi ?

File: Assets/Scripts/Utils/StringHelper.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                else
+                {
+                    description = description.Replace("{" + matches[i].Value + "}", "<b>{" + matches[i].Value + "}</b>");
+                }
+            }
+
+            description = string.Format(description, descriptionVariables);
+        }
+
+        pattern = @"(?<=<<).*?(?=>>)";
```
Comment (Veltouille) [inline Assets/Scripts/Utils/StringHelper.cs:53]: ### Question

Il faudra que tu m'explique comment et sur quoi tu t'es basé pour faire ton pattern. S'il te plait xD

File: Assets/Scripts/Utils/StringHelper.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        if (seconds < 60)
+        {
+            return seconds + " sec.";
+        }
+        else
+        {
+            return string.Format("{0:##.#} min.", (seconds / 60.0));
+        }
+    }
+}
```
Comment (Veltouille) [inline Assets/Scripts/Utils/StringHelper.cs:85]: ### Remarque

J'ai compris à quoi sert la classe. C'est un super travail que tu as fait pour cette static.

Cependant (et parce que je suis chiant), les méthodes de la classes ne sont pas commentées. Je pense que ça pourrait aider parce que c'est pas évident au premier abord. De plus, je pense que tu utilises des codes que tu as fait par toi même et je pense de manière arbitraire (si je ne dit pas de bêtises ^^). Il serait bien de les explicités pour que les autres puisses au plus facilement construire leur propres spells.

Sinon c'est propre :+1:

File: Assets/Scripts/GUI/GUISpellDisplayer.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	 * @Params : Spell
+	 * This public method should only be called by the Character script so far.
+	 * In this method, we attribute to the Gui the correct spell.
+	 * Then, we try to locate a Sprite associated to the spell in the Image Folder associated to the Character.
+	 **/
+    public void AttributeSpellToGUI(Spell spell)
+    {
+        _spell = spell;
+        Image spellImage = GetComponent<Image>();
+        spellImage.sprite = Resources.Load<Sprite>("Images/Spells/" + _spell.GetComponent<Character>().GetType().ToString() + "/" + _spell.GetType());
```
Comment (Veltouille) [inline Assets/Scripts/GUI/GUISpellDisplayer.cs:66]: ### Question

Spell.GetType() te donne le nom du spell ?

File: Assets/Scripts/GUI/GUISpellDisplayer.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    /** Update private void Method
+	 * We ensure that _spellAvailableForGUI is always equal, to the AvailableForGUI() method in the Spell associated.
+	 * We also ensure that the fillAmount of _CDSpellImage is always equal to the current CD of the spell associated. 
+	 * Please note that the fillAmount must be [0;1] this is why we divide the currentCD by the SpellCD.
+	 **/
+    private void Update()
+    {
+        _spellAvailableForGUI.enabled = !_spell.AvailableForGUI();
+        _CDSpellImage.fillAmount = _spell.currentCD / _spell.spellCD;
+        _timerCDText.text = ((int)_spell.currentCD + 1).ToString();
```
Comment (Veltouille) [inline Assets/Scripts/GUI/GUISpellDisplayer.cs:48]: ### Question

_timerCDText.text c'est le petit texte qui est en bas à droite de l'icone de spell.
Mais je ne comprends pas le +1

??? xD

File: Assets/Scripts/Classes/BASE_CHARS/Spell.cs [hunk 7]
```diff
@@ -91,4 +114,76 @@ protected void DisplaySpellCreation(Spell spell)
...
+        public bool HasGCD;
+        public int BaseDamage;
+        public int[] AdditionalDamages;
+        public IStatus[] AdditionalEffects;
+        public string[] OtherValues;
+        public bool IsStackable;
+        public int NumberOfStack;
+        public string[] Description;
+    }
+}
```
Comment (Veltouille) [inline Assets/Scripts/Classes/BASE_CHARS/Spell.cs:189]: Je pense avoir répondu à ma question sur les JSON : Il faut les deux !

L'un pour charger le champion, l'autre pour charger les spells.

Sinon j'ai pas de question sur cette classe, tout est clair.

File: Assets/Scripts/Classes/BASE_CHARS/Character.cs [hunk 8]
```diff
@@ -139,22 +141,36 @@ protected virtual void AttributeAutoAttackToClass()
 	 * This method is called by the Start method. The Objective of the method is to get the spell names in the characterData instance.
 	 * Then, it get all of the script in the scripts library and attach it to the player.
 	 * If one of the scripts is not found or mispelled, the HandleException(3) is launched.
+	 * After that, we call the AttributeSpellToGUI method to give to the GUI all information in requires to display informations about each spell.
 	 **/
     protected virtual void AttributeSpellsToClass()
     {
-        Type t;
-        foreach (string SpellName in characterData.ActiveSpells)
+        for (int i = 1; i <= 4; i++)
```
Comment (Veltouille) [inline Assets/Scripts/Classes/BASE_CHARS/Character.cs:148]: ### Question 

AttributeSpellToGUI(spellToAdd, i); ne peut pas prendre la valeur 0 ?

File: Assets/Scripts/Classes/BASE_CHARS/Character.cs [hunk 9]
```diff
@@ -173,16 +189,17 @@ private void HandleException(int e)
...
-	}
+    [System.Serializable]
+    protected class CharacterData
+    {
+        public string Name;
+        public string Passive;
+        public string AutoAttack;
+        public string[] ActiveSpells;
+    }
 }
```
Comment (Veltouille) [inline Assets/Scripts/Classes/BASE_CHARS/Character.cs:205]: ### Remarque

A part peut être une petite explication sur indexSpell de la méthode AttributeSpellToGUI(Spell spell, int indexSpell).
Je ne vois pas pourquoi il faut que cela commence à 1.

Sinon Nickel t'es toujours un champion =)

**Review by Veltouille — COMMENTED** [review]
Voila je te laisse une bonne série de commets =)

Sinon c'est un sacré truc que tu as pondu là. Toujours admiratif du boulot que tu fais =D

Je te dit merde aussi pour mon travail sur la migration que tu as encore modifié, mais bon ça a heureusement du sens alors j'approuve.

Super boulot Jo, réponds au questions et remarques et je validerai =) :+1:

File: Assets/Scripts/Classes/BASE_CHARS/Character.cs (hunk 8 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/BASE_CHARS/Character.cs:148]: Il le peut c'est simplement que sur le GUI le premier prefab qui accueille le spell s'appelle Spell1 
Du coup faudrait modifier le prefab si on veut le changer.

File: Assets/Scripts/Classes/BASE_CHARS/Character.cs (hunk 9 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/BASE_CHARS/Character.cs:205]: Merci ;)

File: Assets/Scripts/Classes/BASE_CHARS/Spell.cs (hunk 7 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/BASE_CHARS/Spell.cs:189]: Exact à terme, on devrait avoir 
1 Json Pour champion, 1 pour Spells, 1 pour les Autoattacks, 1 pour les passifs et 1 pour le status

Comment (Veltouille) [reply Assets/Scripts/Classes/BASE_CHARS/Spell.cs:189]: Logique Ok !

File: Assets/Scripts/GUI/GUISpellDisplayer.cs (hunk 6 above)
Comment (Myrilandel) [reply Assets/Scripts/GUI/GUISpellDisplayer.cs:48]: Oui c'est ça en fait le +1 est pour une simple question de rendu et d'arrondis. 
Exemple, si mon spell à un CD de 3 sec, au premier tick il passera a 2.98 sec. A l'arrondi à l'écran, on verra 2, puis 1 , puis 0.
Le fait d'ajouter +1 permettra de rendre à l'écran 3 puis 2 puis 1. les deux choix se valent on parlera du rendu ce soir pendant le weekly si tu veux.

File: Assets/Scripts/GUI/GUISpellDisplayer.cs (hunk 5 above)
Comment (Myrilandel) [reply Assets/Scripts/GUI/GUISpellDisplayer.cs:66]: Exact, faut juste lui accoler un ToString pour s'en servir

File: Assets/Scripts/Utils/StringHelper.cs (hunk 2 above)
Comment (Myrilandel) [reply Assets/Scripts/Utils/StringHelper.cs]: C'est une erreur effectivement le P ne saute pas alors qu'il devrait. J'ai fais une refacto hier soir sur ce script et j'ai pas testé avec un P
Bien vu ! je corrige ca ce soir en vitesse

File: Assets/Scripts/GUI/GUISpellDisplayer.cs (hunk 6 above)
Comment (Veltouille) [reply Assets/Scripts/GUI/GUISpellDisplayer.cs:48]: T'es un génie !

File: Assets/Scripts/Utils/StringHelper.cs (hunk 2 above)
Comment (Veltouille) [reply Assets/Scripts/Utils/StringHelper.cs]: ;)

File: Assets/Scripts/Utils/StringHelper.cs (hunk 3 above)
Comment (Myrilandel) [reply Assets/Scripts/Utils/StringHelper.cs:53]: C'est une sorte de printf personnalisé. 
{p0} = La première valeur de la chaîne sera un dégât physique (Gras + Maron).
{m1} = La seconde valeur de la chaîne sera une dégat magique (Gras + Cyan).
{2} = La troisième valeur est une variable ni physique ni magique (Gras).
<<MonTexte>> = Le texte entre <<>> doit représenter le nom d'un autre script à description. (Gras + Vert + Ajoute une sous-section de description) 
Ce sera une façon de dire, j'ajoute (par exemple) <<IgniteStatus>> dans mon text, donc forcément, en dessous de ma description principale, j'ajouterai la description de l'IgniteStatus

File: Assets/Scripts/Utils/StringHelper.cs (hunk 4 above)
Comment (Myrilandel) [reply Assets/Scripts/Utils/StringHelper.cs:85]: Effectivement j'ai zappé les comments de cette méthode alors qu'ils sont absolument nécessaires pour construire une bonne chaîne. Sera ajouté dans le commit de ce soir :)

File: Assets/Scripts/Utils/StringHelper.cs (hunk 3 above)
Comment (Veltouille) [reply Assets/Scripts/Utils/StringHelper.cs:53]: Cela me paraissait important de l'expliciter !

File: Assets/StreamingAssets/SpellData.json (hunk 1 above)
Comment (Myrilandel) [reply Assets/StreamingAssets/SpellData.json:77]: Il faut garder les deux (C.F. Mon commentaires précédent)

Comment (Veltouille) [reply Assets/StreamingAssets/SpellData.json:77]: Je sais, j'ai reviewé par le bas mdr !

**Comment by Veltouille** [comment]
Ok suite à tes réponses je laisse la PR en request changes pour le commit de ce soir ;)

**Comment by Myrilandel** [comment]
Merci pour tous tes commentaires, il y a effectivement des erreurs qu'on ne peut pas laisser partir en prod si tu peux me laisser la PR en request changes stp :) 

Merci beaucoup !

**Review by Veltouille — CHANGES_REQUESTED** [review]
Pour le Commit de ce soir, pour verrouiller en attendant !

File: Assets/Scripts/GUI/GUISpellDisplayer.cs (hunk 6 above)
Comment (Myrilandel) [reply Assets/Scripts/GUI/GUISpellDisplayer.cs:48]: Merci ! :D

**Review by Veltouille — APPROVED** [review]
Un bug fixe de qualité sur ta PR je ne peux que validé ^^

**Comment by Veltouille** [comment]
Aller PR terminée, je merge =)

Super travail Jo, ça va être génial :+1:

**Pull request merged**